README.md
# moxa-push-button

`moxa-push-button` is a C library for using SW push button.

## Build

This project use autotools as buildsystem. You can build this project by the following commands:

* If the build target architecture is x86_64

	```
	# ./autogen.sh --host=x86_64-linux-gnu --includedir=/usr/include/moxa --libdir=/usr/lib/x86_64-linux-gnu --sbindir=/sbin
	# make
	# make install
	```
* If the build target architecture is armhf

	```
	# ./autogen.sh --host=arm-linux-gnueabihf --includedir=/usr/include/moxa --libdir=/usr/lib/arm-linux-gnueabihf --sbindir=/sbin
	# make
	# make install
	```

The autogen script will execute ./configure and pass all the command-line
arguments to it.

## Documentation

[Config Example](/Config_Example.md)

[API Reference](/API_References.md)

debian/changelog
moxa-push-button (1.1.3+deb9) stretch; urgency=low

  * Remove mx_errmsg.
  * Change install directory for libmx_uart_ctl.so*
    from `/lib/<arch>` to `/usr/lib/<arch>`.

 -- Ken CJ Chou <KenCJ.Chou@moxa.com>  Wed, 3 Jul 2019 17:44:00 +0800

moxa-push-button (1.1.2+deb9) stretch; urgency=low

  * fix: use debian/service instead of directly install
  * fix: adjust build-dependencies and dependencies in debian/control
  * fix: change package architecture to "linux-any"

 -- Ken CJ Chou <KenCJ.Chou@moxa.com>  Wed, 22 Aug 2018 14:14:00 +0800

moxa-push-button (1.1.1+deb9) stretch; urgency=low

  * fix: set conditional action to configs

 -- Ken CJ Chou <KenCJ.Chou@moxa.com>  Fri, 17 Aug 2018 18:07:00 +0800

moxa-push-button (1.1.0+deb9) stretch; urgency=low

  * Remove old version of mx-pbtn daemon and LED library codes.
  * Add new mx_pbtn library codes and new moxa-pbtnd daemon code.
  * Redefine deb packaging

 -- Ken CJ Chou <KenCJ.Chou@moxa.com>  Sat, 24 Feb 2018 14:30:00 +0800

moxa-push-button (1.0.0+deb9) stretch; urgency=low

  [ Fero JD Zhou ]
  * Initial release.

  [ SZ Lin (林上智) ]
  * Rename this package

 -- SZ Lin (林上智) <sz.lin@moxa.com>  Sat, 23 Dec 2017 14:32:38 +0800


example/pushbtn-test.c
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Name:
 *	C Example code for MOXA Push button Library
 *
 * Description:
 *	Example code for demonstrating the usage of
 *	MOXA Push button Library in C
 *
 * Authors:
 *	2018	Ken CJ Chou	<KenCJ.Chou@moxa.com>
 */

#include <stdio.h>
#include <stdlib.h>
#include <moxa/mx_pbtn.h>

void pressed_func(int sec)
{
	printf("  This is in pressed_func, sec: %d\n", sec);
}

void released_func(int sec)
{
	printf("  This is in released_func, sec: %d\n", sec);
}

void hold_func(int sec)
{
	printf("  This is in hold_func, sec: %d\n", sec);
}

int main(int argc, char *argv[])
{
	int ret, btn_id;
	unsigned long hold_sec;

	if (argc != 2) {
		fprintf(stderr, "Usage: %s <hold_seconds>\n", argv[0]);
		fprintf(stderr, "	0: function triggered every second\n");
		fprintf(stderr, "	1 ~ 3600: function triggered over hold_seconds\n");
		exit(99);
	}
	hold_sec = atol(argv[1]);

	ret = mx_pbtn_init();
	if (ret < 0) {
		fprintf(stderr, "Error: Initialize Moxa pbtn control library failed\n");
		fprintf(stderr, "Return code: %d\n", ret);
		exit(1);
	}

	printf("Testing Push button:\n");
	printf("========================================\n");

	btn_id = mx_pbtn_open(BUTTON_TYPE_SYSTEM, 1);
	if (btn_id < 0) {
		fprintf(stderr, "Error: Open button failed\n");
		fprintf(stderr, "Return code: %d\n", btn_id);
		exit(1);
	}

	ret = mx_pbtn_pressed_event(btn_id, &pressed_func);
	if (ret < 0) {
		fprintf(stderr, "Error: Set pressed function failed\n");
		fprintf(stderr, "Return code: %d\n", ret);
		exit(1);
	}

	ret = mx_pbtn_released_event(btn_id, &released_func);
	if (ret < 0) {
		fprintf(stderr, "Error: Set released function failed\n");
		fprintf(stderr, "Return code: %d\n", ret);
		exit(1);
	}

	ret = mx_pbtn_hold_event(btn_id, &hold_func, hold_sec);
	if (ret < 0) {
		fprintf(stderr, "Error: Set hold function failed\n");
		fprintf(stderr, "Return code: %d\n", ret);
		exit(1);
	}

	printf("- Start testing. Press button to check the functions.\n");
	printf("  [Ctrl + C] to terminate this process.\n");

	mx_pbtn_wait();

	mx_pbtn_close(btn_id);
	exit(0);
}


mx_led.c
/*##############################################################################
# Copyright (C) MOXA Inc. All rights reserved.
#
# This software is distributed under the terms of the
# MOXA License.  See the file COPYING-MOXA for details.
#
# Purpose : MOXA LED library
# Return : 0 on success; Other wise, modification failed.
# Author : Lock Lin
# Date : 2014-03-17
# Version : 1.0
# Notes :
##############################################################################*/

#include <stdint.h>
#include <linux/input.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <signal.h>
#define MAXLINE 128

#define DEVICE_NODE	"/dev/input/event0"

int mx_btn_event_detect(int fd, int sec)
{
	struct timeval timeout;
	fd_set read_fds;

	FD_ZERO(&read_fds);
	FD_SET(fd, &read_fds);

	timeout.tv_sec = sec;
	timeout.tv_usec = 0;

	return select(fd + 1, &read_fds, NULL, NULL, &timeout);
}

int onoff_led(char *group, char *idx, char *status)
{
	FILE *fp;
	int rt_val;
	char cmd[MAXLINE];

	memset(cmd, '\0', MAXLINE);
	snprintf(cmd, sizeof(cmd), "/sbin/mx-led-ctl -p %s -i %s %s",
group, idx, status);

	fp = popen(cmd, "r");
	if (NULL == fp) {
		perror("Cannot execute cmd\n");
		return -1;
	}
	rt_val = pclose(fp);
	return rt_val;
}

int turn_all_led(char *state)
{
	FILE *fp;
	int rt_val;
	char cmd[MAXLINE];

	memset(cmd, '\0', MAXLINE);
	snprintf(cmd, sizeof(cmd), "/sbin/mx-led-ctl --all-programmable %s",
state);

	fp = popen(cmd, "r");
	if (NULL == fp) {
		perror("Cannot execute cmd\n");
		return -1;
	}
	rt_val = pclose(fp);
	return rt_val;
}

void exec_cmd(char *bin_name)
{
	system(bin_name);
#if 0
	char **cmd[32];
	FILE *fp;
	char rt_val[32];
	char cmd[MAXLINE];

	memset(cmd, '\0', MAXLINE);
	memset(rt_val, '\0', sizeof(rt_val));
	snprintf(cmd, sizeof(cmd), "%s", bin_name);
	fp = popen(cmd, "r");
	printf("cmd=%s\n", cmd);
	if (NULL == fp) {
		perror("Cannot execute cmd\n");
		exit(1);
	}
	fgets(rt_val, 32, fp);
	pclose(fp);
	printf("Exit\n");
#endif
}

int mx_open_device(void)
{
	return open(DEVICE_NODE, O_RDONLY);
}

int mx_close_device(int fd)
{
	close(fd);
}


mx_pbtn.h
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Name:
 *	MOXA Push button Library
 *
 * Description:
 *	Library for setting Push button to trigger actions on input events.
 *
 * Authors:
 *	2014	SZ Lin		<sz.lin@moxa.com>
 *	2017	Fero JD Zhou	<FeroJD.Zhou@moxa.com>
 *	2018	Ken CJ Chou	<KenCJ.Chou@moxa.com>
 */

#ifndef _MOXA_PBTN_H
#define _MOXA_PBTN_H

enum button_type {
	BUTTON_TYPE_SYSTEM = 0,
	BUTTON_TYPE_USER = 1
};

#define DURATION_EVERY 0

#ifdef __cplusplus
extern "C" {
#endif

extern int mx_pbtn_init(void);
extern int mx_pbtn_open(int type, int index);
extern int mx_pbtn_close(int btn_id);
extern int mx_pbtn_wait(void);
extern int mx_pbtn_is_pressed(int btn_id);
extern int mx_pbtn_pressed_event(int btn_id, void (*func)(int));
extern int mx_pbtn_released_event(int btn_id, void (*func)(int));
extern int mx_pbtn_hold_event(int btn_id, void (*func)(int), unsigned long duration);

#ifdef __cplusplus
}
#endif

#endif /* _MOXA_PBTN_H */



push_btn.c
/*
 * Copyright (C) MOXA Inc. All rights reserved.
 *
 * This software is distributed under the terms of the
 * Apache License.
 *
 * Purpose : Push button device
 * Return : 0 on success; Other wise, modification failed.
 * Author : Lock Lin
 * Date : 2014-03-17
 * Version : 1.0
 * Notes :
 * 2016-2017 Fero JD Zhou <FeroJD.Zhou@moxa.com>
 */

#include <stdint.h>
#include <linux/input.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <signal.h>
#include <json-c/json.h>

#ifdef DEBUG
#define DEBUG_PRINT(fmt, args...)    fprintf(stderr, fmt, ## args)
#else
#define DEBUG_PRINT(fmt, args...)    /* Don't do anything in release builds */
#endif

#define CONFIG_FILE "/etc/moxa-configs/moxa-push-button.json"

struct action {
	int interval;
	const char *led_group;
	const char *led_idx;
	const char *led_state;
	const char *debug_msg;
	const char *cmd;
	struct action *release;
};

const char *get_string_from_obj(struct json_object *obj, char *key)
{
	struct json_object *val;

	json_object_object_get_ex(obj, key, &val);
	return json_object_get_string(val);
}

void load_action_from_object(struct action *ac, struct json_object *obj)
{
	ac->interval = atoi(get_string_from_obj(obj, "INTERVAL"));
	ac->led_group = get_string_from_obj(obj, "LED_GROUP");
	ac->led_idx = get_string_from_obj(obj, "LED_IDX");
	ac->led_state = get_string_from_obj(obj, "LED_STATE");
	ac->debug_msg = get_string_from_obj(obj, "DEBUG_MSG");
	ac->cmd = get_string_from_obj(obj, "EXEC_CMD");
}

struct json_object *get_json_obj_from_key(struct json_object *obj, char *key)
{
	struct json_object *val;

	if (!json_object_object_get_ex(obj, key, &val)) {
		printf("Parsing error at key:%s\n", key);
		exit(-1);
	}
	return val;
}

void load_configuration(struct action **list, char *config_name, int *len)
{
	struct json_object *obj = json_object_from_file(config_name);
	struct array_list *arr;
	int i;

	obj = get_json_obj_from_key(obj, "BUTTON1");
	obj = get_json_obj_from_key(obj, "ACTION");
	arr = json_object_get_array(obj);
	*len = array_list_length(arr);
	*list = (struct action *)malloc(sizeof(struct action)*(*len));
	for (i = 0; i < *len; i++) {
		obj = array_list_get_idx(arr, i);
		load_action_from_object(&(*list)[i], obj);
		obj = get_json_obj_from_key(obj, "RELEASE");
		(*list)[i].release =
(struct action *)malloc(sizeof(struct action));
		load_action_from_object((*list)[i].release, obj);
	}
}

static void action_unit_handler(struct action *ac)
{
	if (strcmp(ac->led_idx, "all") == 0)
		turn_all_led(ac->led_state);
	else
		onoff_led(ac->led_group, ac->led_idx, ac->led_state);
}

static void action_handler(int fd, struct action *list, int len)
{
	/*FIXME: We are not sure whether the json_object handles self-destruct*/
	/*automatically or not. May need to run valgrind to check.*/
	struct action *cur_action;
	int i;

	for (i = 0; i < len; i++) {
		turn_all_led("off");
		cur_action = &list[i];
		printf("debug message: %s\n", cur_action->debug_msg);
		action_unit_handler(cur_action);
		if (mx_btn_event_detect(fd, cur_action->interval) == 1) {
			printf("enter during mode(button released!!)\n");
			cur_action = list[i].release;
			action_unit_handler(cur_action);
			exec_cmd(cur_action->cmd);
			break;
		} else {
			printf("enter outside mode\n");
		}
	}
}

int main(int argc, char **argv)
{
	int i, fd, ret, rd, len = 0;
	struct input_event ev[64];
	struct action *list;

	memset(ev, 0, sizeof(struct input_event)*64);
	fd = mx_open_device();

	if (fd < 0) {
		perror("mx-pbtn");
		return 1;
	}

	load_configuration(&list, CONFIG_FILE, &len);

	turn_all_led("off");
	DEBUG_PRINT("Testing ... (interrupt to exit)\n");

	while (1) {
		rd = read(fd, ev, sizeof(struct input_event) * 64);

		if (rd < (int) sizeof(struct input_event)) {
			perror("\nmx-pbtn: error reading");
			return 1;
		}

		if (ev[0].value == 0)
			continue;
		else {
			turn_all_led("off");
			action_handler(fd, list, len);
		}
	}

	for (i = 0; i < len; i++)
		free(list[i].release);

	free(list);
	mx_close_device(fd);
}



tools/moxa-pbtnd.c
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Name:
 *	MOXA Push button Daemon
 *
 * Description:
 *	Daemon that loads default actions for system type push button.
 *
 * Authors:
 *	2014	SZ Lin		<sz.lin@moxa.com>
 *	2017	Fero JD Zhou	<FeroJD.Zhou@moxa.com>
 *	2018	Ken CJ Chou	<KenCJ.Chou@moxa.com>
 */

#include <stdio.h>
#include <string.h>
#include <json-c/json.h>
#include <moxa/mx_led.h>
#include <mx_pbtn.h>

#define CONF_FILE "/etc/moxa-configs/moxa-push-button.json"

static struct json_object *config;

/*
 * json-c utilities
 */

static inline int obj_get_obj(struct json_object *obj, char *key, struct json_object **val)
{
	if (!json_object_object_get_ex(obj, key, val))
		return -1;
	return 0;
}

static int obj_get_int(struct json_object *obj, char *key, int *val)
{
	struct json_object *tmp;

	if (obj_get_obj(obj, key, &tmp) < 0)
		return -1;

	*val = json_object_get_int(tmp);
	return 0;
}

static int obj_get_str(struct json_object *obj, char *key, const char **val)
{
	struct json_object *tmp;

	if (obj_get_obj(obj, key, &tmp) < 0)
		return -1;

	*val = json_object_get_string(tmp);
	return 0;
}

static int obj_get_arr(struct json_object *obj, char *key, struct array_list **val)
{
	struct json_object *tmp;

	if (obj_get_obj(obj, key, &tmp) < 0)
		return -1;

	*val = json_object_get_array(tmp);
	return 0;
}

static int arr_get_obj(struct array_list *arr, int idx, struct json_object **val)
{
	if (arr == NULL || idx >= arr->length)
		return -1;

	*val = array_list_get_idx(arr, idx);
	return 0;
}

static int arr_get_int(struct array_list *arr, int idx, int *val)
{
	struct json_object *tmp;

	if (arr_get_obj(arr, idx, &tmp) < 0)
		return -1;

	*val = json_object_get_int(tmp);
	return 0;
}

static int arr_get_str(struct array_list *arr, int idx, const char **val)
{
	struct json_object *tmp;

	if (arr_get_obj(arr, idx, &tmp) < 0)
		return -1;

	*val = json_object_get_string(tmp);
	return 0;
}

static int arr_get_arr(struct array_list *arr, int idx, struct array_list **val)
{
	struct json_object *tmp;

	if (arr_get_obj(arr, idx, &tmp) < 0)
		return -1;

	*val = json_object_get_array(tmp);
	return 0;
}

/*
 * main part
 */

static void init(void)
{
	struct json_object *obj;
	struct array_list *arr;

	obj = json_object_from_file(CONF_FILE);
	if (obj == NULL) {
		fprintf(stderr, "json-c: load file %s failed\n", CONF_FILE);
		exit(1);
	}

	if (obj_get_arr(obj, "DEFAULT_ACTIONS", &arr) < 0)
		exit(1);

	if (arr_get_obj(arr, 0, &config) < 0)
		exit(1);
}

void do_action(struct json_object *action)
{
	int group, index;
	const char *state, *message, *cmd;

	if ((obj_get_int(action, "LED_GROUP", &group) == 0) &&
		(obj_get_int(action, "LED_INDEX", &index) == 0) &&
		(obj_get_str(action, "LED_STATE", &state) == 0)) {
		if (mx_led_set_type_all(LED_TYPE_PROGRAMMABLE, LED_STATE_OFF) < 0)
			exit(2);

		if (strcmp(state, "off") == 0) {
			if (mx_led_set_brightness(LED_TYPE_PROGRAMMABLE,
				group, index, LED_STATE_OFF) < 0)
				exit(2);
		} else if (strcmp(state, "on") == 0) {
			if (mx_led_set_brightness(LED_TYPE_PROGRAMMABLE,
				group, index, LED_STATE_ON) < 0)
				exit(2);
		} else if (strcmp(state, "blink") == 0) {
			if (mx_led_set_brightness(LED_TYPE_PROGRAMMABLE,
				group, index, LED_STATE_BLINK) < 0)
				exit(2);
		}
	}

	if (obj_get_str(action, "MESSAGE", &message) == 0)
		printf("%s\n", message);

	if (obj_get_str(action, "EXEC_CMD", &cmd) == 0)
		system(cmd);
}

void pressed_func(int sec)
{
	struct array_list *actions;
	struct json_object *action;

	if (obj_get_arr(config, "PRESS_ACTION", &actions) < 0)
		exit(1);

	if (arr_get_obj(actions, 0, &action) < 0)
		exit(1);

	do_action(action);
}

void released_func(int sec)
{
	struct array_list *actions;
	struct json_object *action;
	int i, t;

	if (obj_get_arr(config, "RELEASE_ACTION", &actions) < 0)
		exit(1);
	for (i = 0; i < array_list_length(actions); i++) {
		if (arr_get_obj(actions, i, &action) < 0)
			exit(1);

		if (obj_get_int(action, "SEC", &t) < 0)
			exit(1);

		if (sec >= t) {
			do_action(action);
			break;
		}
	}
}

void hold_func(int sec)
{
	struct array_list *actions;
	struct json_object *action;
	int i, t;

	if (obj_get_arr(config, "HOLD_ACTION", &actions) < 0)
		exit(1);
	for (i = 0; i < array_list_length(actions); i++) {
		if (arr_get_obj(actions, i, &action) < 0)
			exit(1);

		if (obj_get_int(action, "SEC", &t) < 0)
			exit(1);

		if (sec > t) {
			break;
		} else if (sec == t) {
			do_action(action);
			break;
		}
	}
}

int main(int argc, char *argv[])
{
	int btn_id;

	init();
	if (mx_pbtn_init() < 0) {
		fprintf(stderr, "Initialize Moxa push button library failed\n");
		exit(3);
	}
	if (mx_led_init() < 0) {
		fprintf(stderr, "Initialize Moxa LED control library failed\n");
		exit(2);
	}

	btn_id = mx_pbtn_open(BUTTON_TYPE_SYSTEM, 1);
	if (btn_id < 0)
		exit(3);
	if (mx_pbtn_pressed_event(btn_id, &pressed_func) < 0)
		exit(3);
	if (mx_pbtn_released_event(btn_id, &released_func) < 0)
		exit(3);
	if (mx_pbtn_hold_event(btn_id, &hold_func, DURATION_EVERY) < 0)
		exit(3);

	if (mx_pbtn_wait() < 0)
		exit(3);

	if (mx_pbtn_close(btn_id) < 0)
		exit(3);

	exit(0);
}


lib/mx_pbtn.c
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Name:
 *	MOXA Push button Library
 *
 * Description:
 *	Library for setting Push button to trigger actions on input events.
 *
 * Authors:
 *	2014	SZ Lin		<sz.lin@moxa.com>
 *	2017	Fero JD Zhou	<FeroJD.Zhou@moxa.com>
 *	2018	Ken CJ Chou	<KenCJ.Chou@moxa.com>
 */

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <pthread.h>
#include <errno.h>
#include <sys/file.h>
#include <linux/input.h>
#include <json-c/json.h>
#include <mx_pbtn.h>

#define CONF_FILE "/etc/moxa-configs/moxa-push-button.json"
#define CONF_VER_SUPPORTED "1.1.*"
#define MAX_FILEPATH_LEN 256	/* reserved length for file path */

struct button_struct {
	int fd;
	pthread_t thread;
	int is_opened;
	int is_pressed;
	void (*pressed_func)(int sec);
	void (*released_func)(int sec);
	void (*hold_func)(int sec);
	int hold_duration;
};

static int lib_initialized;
static struct json_object *config;
static struct button_struct *buttons;

/*
 * json-c utilities
 */

static inline int obj_get_obj(struct json_object *obj, char *key, struct json_object **val)
{
	if (!json_object_object_get_ex(obj, key, val))
		return -1;
	return 0;
}

static int obj_get_int(struct json_object *obj, char *key, int *val)
{
	struct json_object *tmp;

	if (obj_get_obj(obj, key, &tmp) < 0)
		return -1;

	*val = json_object_get_int(tmp);
	return 0;
}

static int obj_get_str(struct json_object *obj, char *key, const char **val)
{
	struct json_object *tmp;

	if (obj_get_obj(obj, key, &tmp) < 0)
		return -1;

	*val = json_object_get_string(tmp);
	return 0;
}

static int obj_get_arr(struct json_object *obj, char *key, struct array_list **val)
{
	struct json_object *tmp;

	if (obj_get_obj(obj, key, &tmp) < 0)
		return -1;

	*val = json_object_get_array(tmp);
	return 0;
}

static int arr_get_obj(struct array_list *arr, int idx, struct json_object **val)
{
	if (arr == NULL || idx >= arr->length)
		return -1;

	*val = array_list_get_idx(arr, idx);
	return 0;
}

static int arr_get_int(struct array_list *arr, int idx, int *val)
{
	struct json_object *tmp;

	if (arr_get_obj(arr, idx, &tmp) < 0)
		return -1;

	*val = json_object_get_int(tmp);
	return 0;
}

static int arr_get_str(struct array_list *arr, int idx, const char **val)
{
	struct json_object *tmp;

	if (arr_get_obj(arr, idx, &tmp) < 0)
		return -1;

	*val = json_object_get_string(tmp);
	return 0;
}

static int arr_get_arr(struct array_list *arr, int idx, struct array_list **val)
{
	struct json_object *tmp;

	if (arr_get_obj(arr, idx, &tmp) < 0)
		return -1;

	*val = json_object_get_array(tmp);
	return 0;
}

/*
 * static functions
 */

static int check_config_version_supported(const char *conf_ver)
{
	int cv[2], sv[2];

	if (sscanf(conf_ver, "%d.%d.%*s", &cv[0], &cv[1]) < 0)
		return -1; /* E_SYSFUNCERR */

	if (sscanf(CONF_VER_SUPPORTED, "%d.%d.%*s", &sv[0], &sv[1]) < 0)
		return -1; /* E_SYSFUNCERR */

	if (cv[0] != sv[0] || cv[1] != sv[1])
		return -4; /* E_UNSUPCONFVER */
	return 0;
}

static int get_button(int btn_id, struct button_struct **button)
{
	int num_of_all_btns;

	if (obj_get_int(config, "NUM_OF_ALL_BUTTONS", &num_of_all_btns) < 0)
		return -5; /* E_CONFERR */

	if (btn_id < 0 || btn_id > num_of_all_btns - 1)
		return -2; /* E_INVAL */

	*button = &buttons[btn_id];
	return 0;
}

static int detect_input(int fd)
{
	fd_set read_fds;
	struct timeval timeout;

	FD_ZERO(&read_fds);
	FD_SET(fd, &read_fds);

	timeout.tv_sec = 1;
	timeout.tv_usec = 0;

	return select(fd + 1, &read_fds, NULL, NULL, &timeout);
}

static void handle_event(struct button_struct *button)
{
	int sec = 0;

	if (button->pressed_func != NULL)
		button->pressed_func(0);

	while (1) {
		sec++;

		if (detect_input(button->fd) == 1) {
			button->is_pressed = 0;
			if (button->released_func != NULL)
				button->released_func(sec - 1);
			break;
		}

		if (button->hold_func != NULL) {
			if (button->hold_duration == DURATION_EVERY)
				button->hold_func(sec);
			else if (sec == button->hold_duration)
				button->hold_func(sec);
		}
	}
}

static void *thread_start(void *arg)
{
	struct button_struct *button = (struct button_struct *) arg;
	struct input_event events[64];
	int read_len;

	memset(events, 0, sizeof(struct input_event) * 64);

	while (1) {
		read_len = read(button->fd, events
			, sizeof(struct input_event) * 64);

		if (read_len < (int) sizeof(struct input_event)) {
			perror("read input error");
			/* to do: error handling */
			return NULL;
		}

		if (events[0].value == 0) {
			continue;
		} else {
			button->is_pressed = 1;
			handle_event(button);
		}
	}
	return NULL;
}

/*
 * APIs
 */

int mx_pbtn_init(void)
{
	int ret, num_of_all_btns, i;
	const char *conf_ver;

	if (lib_initialized)
		return 0;

	config = json_object_from_file(CONF_FILE);
	if (config == NULL)
		return -5; /* E_CONFERR */

	if (obj_get_str(config, "CONFIG_VERSION", &conf_ver) < 0)
		return -5; /* E_CONFERR */

	ret = check_config_version_supported(conf_ver);
	if (ret < 0)
		return ret;

	if (obj_get_int(config, "NUM_OF_ALL_BUTTONS", &num_of_all_btns) < 0)
		return -5; /* E_CONFERR */

	buttons = (struct button_struct *)
		malloc(num_of_all_btns * sizeof(struct button_struct));
	if (buttons == NULL)
		return -1; /* E_SYSFUNCERR */

	for (i = 0; i < num_of_all_btns; i++) {
		buttons[i].is_opened = 0;
		buttons[i].pressed_func = NULL;
		buttons[i].released_func = NULL;
		buttons[i].hold_func = NULL;
	}

	lib_initialized = 1;
	return 0;
}

int mx_pbtn_open(int type, int index)
{
	struct array_list *btn_types, *btn_paths;
	struct json_object *btn_type_info;
	struct button_struct *button;
	const char *filepath;
	int ret, btn_id;

	if (!lib_initialized)
		return -3; /* E_LIBNOTINIT */

	if (obj_get_arr(config, "BUTTON_TYPES", &btn_types) < 0)
		return -5; /* E_CONFERR */

	if (arr_get_obj(btn_types, BUTTON_TYPE_SYSTEM, &btn_type_info) < 0)
		return -5; /* E_CONFERR */

	btn_id = index - 1;
	if (type == BUTTON_TYPE_USER) {
		int sys_btn_num;

		if (obj_get_int(btn_type_info, "NUM_OF_BUTTONS", &sys_btn_num) < 0)
			return -5; /* E_CONFERR */
		btn_id += sys_btn_num;
		if (arr_get_obj(btn_types, BUTTON_TYPE_USER, &btn_type_info) < 0)
			return -5; /* E_CONFERR */
	}

	ret = get_button(btn_id, &button);
	if (ret < 0)
		return ret;

	if (button->is_opened == 1)
		return 0;

	if (obj_get_arr(btn_type_info, "PATHS", &btn_paths) < 0)
		return -5; /* E_CONFERR */

	if (arr_get_str(btn_paths, index - 1, &filepath) < 0)
		return -5; /* E_CONFERR */

	button->fd = open(filepath, O_RDONLY);
	if (button->fd < 0)
		return -1; /* E_SYSFUNCERR */
	flock(button->fd, LOCK_EX);

	pthread_create(&button->thread, NULL, thread_start, button);
	button->is_opened = 1;
	button->is_pressed = 0;
	return btn_id;
}

int mx_pbtn_close(int btn_id)
{
	struct button_struct *button;
	int ret;

	if (!lib_initialized)
		return -3; /* E_LIBNOTINIT */

	ret = get_button(btn_id, &button);
	if (ret < 0)
		return ret;

	if (button->is_opened == 0)
		return 0;

	pthread_cancel(button->thread);
	close(button->fd);
	button->pressed_func = NULL;
	button->released_func = NULL;
	button->hold_func = NULL;
	button->is_opened = 0;
	return 0;
}

int mx_pbtn_wait(void)
{
	struct button_struct *button;
	int ret, btn_num, i = 0;

	if (!lib_initialized)
		return -3; /* E_LIBNOTINIT */

	if (obj_get_int(config, "NUM_OF_ALL_BUTTONS", &btn_num) < 0)
		return -5; /* E_CONFERR */

	while (i < btn_num) {
		ret = get_button(i, &button);
		if (ret < 0)
			return ret;

		if (button->is_opened == 1) {
			pthread_join(button->thread, NULL);
			i = 0;
		} else {
			i++;
		}
	}
	return 0;
}

int mx_pbtn_is_pressed(int btn_id)
{
	struct button_struct *button;
	int ret;

	if (!lib_initialized)
		return -3; /* E_LIBNOTINIT */

	ret = get_button(btn_id, &button);
	if (ret < 0)
		return ret;

	if (button->is_opened != 1)
		return -70; /* E_PBTN_NOTOPEN */

	return button->is_pressed;
}

int mx_pbtn_pressed_event(int btn_id, void (*func)(int))
{
	struct button_struct *button;
	int ret;

	if (!lib_initialized)
		return -3; /* E_LIBNOTINIT */

	ret = get_button(btn_id, &button);
	if (ret < 0)
		return ret;

	button->pressed_func = func;
	return 0;
}

int mx_pbtn_released_event(int btn_id, void (*func)(int))
{
	struct button_struct *button;
	int ret;

	if (!lib_initialized)
		return -3; /* E_LIBNOTINIT */

	ret = get_button(btn_id, &button);
	if (ret < 0)
		return ret;

	button->released_func = func;
	return 0;
}

int mx_pbtn_hold_event(int btn_id, void (*func)(int), unsigned long duration)
{
	struct button_struct *button;
	int ret;

	if (!lib_initialized)
		return -3; /* E_LIBNOTINIT */

	ret = get_button(btn_id, &button);
	if (ret < 0)
		return ret;

	if (duration > 3600)
		return -2; /* E_INVAL */

	button->hold_func = func;
	button->hold_duration = duration;
	return 0;
}
